spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java

```diff
@@ -109,89 +113,251 @@ public Object postProcessAfterInitialization(final Object bean, String beanName)
-		for (int parameterIndex = 0; parameterIndex < arguments.length; parameterIndex++) {
+	private int inputAnnotationCount(Method method) {
+		int inputAnnotationCount = 0;
+		for (int parameterIndex = 0; parameterIndex < method.getParameterTypes().length; parameterIndex++) {
 			MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, parameterIndex);
-			Class<?> parameterType = methodParameter.getParameterType();
-			Annotation targetReferenceAnnotation = methodParameter.hasParameterAnnotation(Input.class) ?
-					methodParameter.getParameterAnnotation(Input.class) : methodParameter.getParameterAnnotation(
-					Output.class);
-			Object targetReferenceAnnotationValue = AnnotationUtils.getValue(targetReferenceAnnotation);
-			Assert.isInstanceOf(String.class, targetReferenceAnnotationValue, "Annotation value must be a String");
-			Assert.hasText((String) targetReferenceAnnotationValue, "Annotation value must not be empty");
-			Object targetBean = this.applicationContext.getBean((String) targetReferenceAnnotationValue);
-			if (parameterType.isAssignableFrom(targetBean.getClass())) {
-				arguments[parameterIndex] = targetBean;
+			if (methodParameter.hasParameterAnnotation(Input.class)) {
+				inputAnnotationCount++;
 			}
-			else {
-				for (StreamListenerParameterAdapter<?, Object> streamListenerParameterAdapter :
-						this.streamListenerParameterAdapters) {
-					if (streamListenerParameterAdapter.supports(targetBean.getClass(), methodParameter)) {
-						arguments[parameterIndex] = streamListenerParameterAdapter.adapt(targetBean, methodParameter);
-						break;
-					}
+		}
+		return inputAnnotationCount;
+	}
+
+	private int outputAnnotationCount(Method method) {
+		int outputAnnotationCount = 0;
+		for (int parameterIndex = 0; parameterIndex < method.getParameterTypes().length; parameterIndex++) {
+			MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, parameterIndex);
+			if (methodParameter.hasParameterAnnotation(Output.class)) {
+				outputAnnotationCount++;
+			}
+		}
+		return outputAnnotationCount;
+	}
+
+	private boolean isDeclarativeStreamListenerMethod(Method method, String methodAnnotatedinboundName, String methodAnnotatedOutboundName) {
+		int methodArgumentsLength = method.getParameterTypes().length;
+		for (int parameterIndex = 0; parameterIndex < methodArgumentsLength; parameterIndex++) {
+			MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, parameterIndex);
+			if (methodParameter.hasParameterAnnotation(Input.class)) {
+				String inboundName = (String) AnnotationUtils.getValue(methodParameter.getParameterAnnotation(Input.class));
+				Assert.isTrue(StringUtils.hasText(inboundName), "@Input annotation should always be associated with a valid inbound name");
+				return isDeclarativeMethodParameter(getBindableBean(inboundName), methodParameter);
+			}
+			else if (methodParameter.hasParameterAnnotation(Output.class)) {
```

Comment:
Since `getBindableBean` is an expensive operation, I thought it is better to return `true` as soon as we find the `declarative` method parameter and so we don't have to loop through all the method parameters and then `return`.